io_scene_godot/converters/material_node_tree.py

```diff
@@ -0,0 +1,1090 @@
+        """Convert to a definition str"""
+        return self.type + ' ' + str(self)
+
+    def __str__(self):
+        """Convert to string"""
+        return self.name
+
+
+class Value:
+    """A constant value in material shader scripts"""
+
+    def __init__(self, type_str, data):
+        self.type = type_str
+        self.data = data
+
+    @classmethod
+    def create_from_blender_value(cls, blender_value):
+        """Creaate a Value() from a blender object"""
+        if isinstance(
+                blender_value, (bpy.types.bpy_prop_array, mathutils.Vector)):
+            tmp = list()
+            for val in blender_value:
+                tmp.append(val)
+
+            return Value("vec{}".format(len(tmp)), tuple(tmp))
+
+        if isinstance(blender_value, mathutils.Matrix):
+            fixed_mat = fix_matrix(blender_value)
+            tmp = list()
+            for vec in mathutils.Matrix:
+                tmp.append(vec)
+
+            return Value("mat{}".format(len(tmp)), tuple(tmp))
+
+        return Value("float", blender_value)
+
+    def __str__(self):
+        """Convert to string"""
+        if self.type.startswith(('vec', 'mat')):
+            return "{}({})".format(
+                self.type,
+                ', '.join([str(x) for x in self.data])
+            )
+        return str(self.data)
+
+
+class FragmentOutChannels:
+    """A tuple of channels represents blender shader node output"""
+    CHANNEL_TYPES = collections.OrderedDict({
+        'albedo': 'vec3',
+        'alpha': 'float',
+        'sss_strength': 'float',
+        'specular': 'float',
+        'metallic': 'float',
+        'roughness': 'float',
+        'clearcoat': 'float',
+        'clearcoat_gloss': 'float',
+        'anisotropy': 'float',
+        'transmission': 'float',
+        'eta': 'float',
+        'emission': 'vec3',
+        'normal': 'vec3',
+        'tangent': 'vec3',
+    })
+
+    def __init__(self):
+        self._data = collections.OrderedDict()
+
+    def get_channel(self, channel_name):
+        """Get a channel value, if the channel is empty return None"""
+        return self._data.get(channel_name, None)
+
+    def set_channel(self, channel_name, channel_value):
+        """Set a channel, note that channel_value can be
+        either Value() or Variable()"""
+        self._data[channel_name] = channel_value
+
+    @classmethod
+    def get_type(cls, channel_name):
+        """Return a type string of a channel"""
+        return cls.CHANNEL_TYPES[channel_name]
+
+
+INV_VIEW_MATRIX = Variable('mat4', 'inverted_view_matrix')
+INV_WORLD_MATRIX = Variable('mat4', 'inverted_world_matrix')
+
+# uniform to easily adjust refraction effect
+REFRACTION_OFFSET = Variable('float', 'refraction_offset')
+
+
+class GlobalScript:
+    """global scripts includes render mode, uniforms and functions"""
+    def __init__(self):
+        # render mode and render type is also
+        # placed here
+        self.uniform_codes = Array(
```

Comment:
Hmm. Curious way of using the Array type. I would have hardcoded the `shader_type spatial` in the to_string method of this class and then used an array for the render_modes: `prefix="render_mode ", separator=", ", suffix=";"`

Reply:
that seems better, more align to the `uniform_codes` name and a more extensible render mode,  thank you for the suggestion